### PR TITLE
[WIP] Add an alternative data type

### PR DESCRIPTION
**Pull request by snowleopard**
This seems like a neat extension :-)

### CODE REVIEW

**Comment by snowleopard** [comment]
Note: I haven't yet checked that composition is associative.

**Comment by snowleopard** [comment]
Interestingly, `Add` is a bit like `Overlay` in algebraic graphs.

**Comment by bolt12** [comment]
Cool! Thanks for the PR :smile: 

As I said this also feels like a LA DSL à la ([Practical Probabilistic Programming with Monads](https://www.repository.cam.ac.uk/bitstream/handle/1810/249132/Scibior%20et%20al%202015%20Haskell%20Symposium%202015.pdf;sequence=1)). Using this structure, LA algorithms can be seen as structure transformations waiting to be interpreted.

Eventually one would need to normalize the structure just to a set of `Join`/`Fork`/`Scale` in order to "evaluate"/"interpret" the structure, right?

**Comment by snowleopard** [comment]
@bolt12 Indeed, this feels like a deep embedding of a DSL built using the language of cartesian/co-cartesian categories -- see my latest commit which adds the corresponding instances, and also `Zero` constructor. 

> Eventually one would need to normalize the structure just to a set of `Join`/`Fork`/`Scale` in order to "evaluate"/"interpret" the structure, right?

Not sure what you mean by "evaluate" -- what is the type signature you have in mind?

**Comment by snowleopard** [comment]
The `Cartesian (Matrix e)` and `CoCartesian (Matrix e)` instances are just awesome :)

**Comment by bolt12** [comment]
Awesome work! Never thought about doing that!

> Not sure what you mean by "evaluate" -- what is the type signature you have in mind?

Well I was just trying to wrap my head around how would `Show` instance be for example..
Imagining applying this data type to a concrete example would lead to (when dealing with large matrices) a lot of room to optimize the internal structure (ex: getting rid of `Add`). 
**Edit:** After thinking about this a little I think that a function `normalize :: Matrix e a b -> Matrix e a b` would make sense. This function would remove all operations (we only have `Add` for now) and return a _pure_ matrix. All the matrices manipulated in terms of this data type are purely syntactic in a way that they do not translate (directly) to a matrix. I think this should be the major difference between this alternative definition and the original one.

How would Khatri Rao product look like? :smile: **Edit:** Probably would need to add another constructor and regarding its projections they might be trickier to have

**Comment by snowleopard** [comment]
@bolt12 Not yet sure about `Show` or `kr`, but we get `><` and `-|-` for free! :)

**Comment by bolt12** [comment]
This is turning out very cool! But just want to point out another reason to not use this structure as the primary "internal" definition of `Matrix`:

I'm trying to implement a `normalize :: Matrix e a b -> Matrix e a b` function as I said above and the lack of an explicit inductive base case makes writing this function a bit awkward. For example

```Haskell
normalize :: Matrix e a b -> Matrix e a b
normalize m = case m of
    -- Normalization
    Scale x Zero                -> Zero
    Scale x Identity            -> _
    Scale x (Join a b)          -> Join (normalize (Scale x a)) (normalize (Scale x b))
    Scale x (Fork a b)          -> Fork (normalize (Scale x a)) (normalize (Scale x b))
    Add Identity Identity       -> _
    Add (Scale x y) Identity    -> _
    Add Identity (Scale x y)    -> _
    Add Zero x                  -> x
    Add x Zero                  -> x
    Add (Scale a x) (Scale b y) -> _
    Add (Join a b) (Join c d)   -> Join (normalize (Add a c)) (normalize (Add b d))
    Add (Fork a b) (Fork c d)   -> Fork (normalize (Add a c)) (normalize (Add b d))
    -- (etc ...)
    -- Distribute normalize
    Zero      -> Zero
    Identity  -> Identity
    Scale x y -> Scale x (normalize y)
    Join a b  -> Join (normalize a) (normalize b)
    Fork a b  -> Fork (normalize a) (normalize b)
    Add a b -> Add (normalize a) (normalize b)
```

The holes are cases that I think are non-trivial when using this new definition because dimensions `a` and `b` are arbitrary and we need to be careful with that. A solution would be to define a class `Normalize e a b` and write instances `Normalize e Void Void` and `Normalize e () ()` to appropriately deal with the base cases but `Normalize e a b` would still not be straightforward to write.

Basically in order to write less simple algorithms using this matrix definition one needs to recur to Type Classes which is not as elegant and trickier (bug prone).

**Edit:** I use `(><)` as the matrix Kronecker product. In LA `(-|-)` is seen as direct sum and **your** `(><)` is just `(><) a b = transpose (a -|- b)`

**Comment by snowleopard** [comment]
@bolt12 Yes, I had a feeling that my definition of `><` might be different from yours. Don't pay too much attention to this experiment, I'm just trying to see where this path leads :)

I've just pushed a minor refactoring, replacing `Scale` and `Add` with a single `Lift`. This allowed me to implement something which looks like `kr`, but probably isn't :)

What is the semantics of your `normalise` function? Does it produce some sort of a canonical matrix representation?

File: src/LAoP/Matrix/Alternative.hs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    Identity   -> Identity
+    Zero       -> Zero
+    Lift f x y -> Lift f (transpose x) (transpose y)
+    Join x y   -> Fork (transpose x) (transpose y)
+    Fork x y   -> Join (transpose x) (transpose y)
+
+select :: Num e => Matrix e a (Either b c) -> Matrix e b c -> Matrix e a c
+select x y = Join y id . x
+
+kr :: Num e => Matrix e a b -> Matrix e a c -> Matrix e a (Either b c)
```
Comment (bolt12) [inline src/LAoP/Matrix/Alternative.hs]: Khatri Rao type should be `kr :: Num e => Matrix e a b -> Matrix e a c -> Matrix e a (b, c)`. Matrix has a Monoidal structure! That's why I said it might be tricky to implement without a constructor.

References: 
- https://pdfs.semanticscholar.org/ccf5/27fa9179081223bffe8067edd81948644fc0.pdf (page 12)
- https://haslab.uminho.pt/jno/files/scp11.pdf (page 38)

**Review by bolt12 — COMMENTED** [review]
> What is the semantics of your `normalise` function? Does it produce some sort of a canonical matrix representation?

`normalise` just gets rid of any unnecessary "syntax", in your other version you had `Add` and `Scale` and `normalise` should remove those constructors and return a "pure" matrix. It doesn't need to be canonical but just a matrix defined in terms of `Join`/`Fork`/`Identity`/`Scale x (Identity :: Matrix e () ()`

File: src/LAoP/Matrix/Alternative.hs (hunk 1 above)
Comment (snowleopard) [reply src/LAoP/Matrix/Alternative.hs]: Oops, you're right! I got carried away replacing `fstM` and `sndM` with their `M`-less versions :)

**Comment by bolt12** [comment]
@snowleopard I think I messed up the previous comment. I'll repeat it here:

> What is the semantics of your `normalise` function? Does it produce some sort of a canonical matrix representation?

`normalise` just gets rid of any unnecessary "syntax", in your other version you had `Add` and `Scale` and `normalise` should remove those constructors and return a "pure" matrix. It doesn't need to be canonical but just a matrix defined in terms of `Join`/`Fork`/`Identity`/`Scale x (Identity :: Matrix e () ()`

File: src/LAoP/Matrix/Alternative.hs (hunk 1 above)
Comment (bolt12) [reply src/LAoP/Matrix/Alternative.hs]: No problem! :smile:  Probably another constructor will solve the issue, but I'm not sure about the projections.

**Comment by snowleopard** [comment]
@bolt12 Thanks, I think I see where the main difficulty is now. Even implementing basic functions like `toLists` is challenging with this alternative data type. I'll keep this PR open, perhaps we'll figure out a good approach :)

**Comment by bolt12** [comment]
Agree! Thanks for your help! :smiley:

**Comment by snowleopard** [comment]
Thanks to an insight of my colleague @aalekseyev, I managed to implement `toLists` :-)

The insight was: define the semantics of the `Matrix e a b` type as a map on vectors and work with it.

Defining the semantics itself is not too difficult, and once you have it, functions like `toLists` can be written in a relatively straightforward way.

```haskell
> m = (one 1 ||| one 2) &&& (one 3 ||| one 4)
> dump m
[1,2]
[3,4]
> dump (transpose m)
[1,3]
[2,4]
> dump (m . transpose m)
[5,11]
[11,25]
> dump ((m . transpose m) ||| id)
[5,11,1,0]
[11,25,0,1]
```

**Comment by bolt12** [comment]
Very interesting approach! :smile: can this be seen as somewhat interpreting the DSL ?

**Comment by snowleopard** [comment]
@bolt12 Well, any algebraic data type can be viewed as a small DSL, and any function operating no this data type -- as an interpreter for this DSL, so I'd say the answer is yes :-)

**Comment by bolt12** [comment]
I see, you're right! Thanks for this solution!

**Comment by snowleopard** [comment]
@bolt12 I'm closing the PR since it has drifted quite far from the library, and there doesn't seem to be any point in merging it.

I've moved the code to a repository I use for experimenting with various CT ideas:

https://github.com/snowleopard/cats/blob/master/src/Matrix.hs

I'll let you know if it evolves in something interesting!
